Refactor ReplaceReferencesService

NobodysNightmare opened the pull request:
Instead of expecting modules to selectively overwrite methods of this class, we offer an explicit extension point where replacements can be registered explicitly.

Thus patching the implementation is not necessary anymore, instead it can be configured from initializers.

# Ticket
none

# What are you trying to accomplish?
Avoid the need to overwrite class behaviour from modules.

# What approach did you choose and why?
Allowing to configure rewrites, because the implementation is kind of always the same. Thus avoiding to duplicate the implementations. This also works nicer with modules than expecting class overwrites.

NobodysNightmare committed:
Refactor ReplaceReferencesService

Instead of expecting modules to selectively overwrite
methods of this class, we offer an explicit extension point
where replacements can be registered explicitly.

Thus patching the implementation is not necessary anymore, instead
it can be configured from initializers.

NobodysNightmare commented:
This will conflict with the changes from https://github.com/opf/openproject/pull/17477 once they have been merged.

Preferably we merge https://github.com/opf/openproject/pull/17477 first, so that I can fix the conflicts in this branch, since the other PR targets release 15.1.

oliverguenther commented:
@NobodysNightmare Other PR was merged

NobodysNightmare commented:
Thanks. Conflicts have been resolved.

machisuji commented on modules/storages/lib/open_project/storages/engine.rb:
```diff
@@ -50,6 +50,14 @@ def self.external_file_permissions
 
+    initializer "openproject_storages.replace_references" do
+      Rails.application.reloader.to_prepare do
+        Principals::ReplaceReferencesService.add_replacement("::Storages::Storage", :creator_id)
```
I suggest adding a method to [ActsAsOpEngine ](https://github.com/opf/openproject/blob/dev/lib/open_project/plugins/acts_as_op_engine.rb) to allow plugins adding these replacements.

NobodysNightmare replied:
Nice catch, I didn't think of this. Will need to wrap my head around it, but sounds like I would love to do that.

NobodysNightmare replied:
Pending tests confirming that I correctly implemented this, I added helpers for these replacements. Thanks for the suggestion.

machisuji commented on config/initializers/replace_references_service.rb:
```diff
@@ -0,0 +1,54 @@
+#++
+
+Rails.application.reloader.to_prepare do
+  Principals::ReplaceReferencesService.add_replacement("AuthProvider", :creator_id)
```
Could you please just put the parameters in a big old hash and only mention `Principals::ReplaceReferencesService` once?

You could probably have the foreign keys as the keys as well to save on repeating those. Just makes it easier to keep track of things and to change things later, in case that's needed.

NobodysNightmare replied:
> Could you please just put the parameters in a big old hash and only mention Principals::ReplaceReferencesService once?

I'll give that a shot, though it means that I will have to add code that can merge the lists/sets. I have to implement it to see, whether I like it :)

> You could probably have the foreign keys as the keys as well to save on repeating those

I already thought about that when sorting the lines in this initializer (i.e. should I group by attribute or by class?). In the end I think it doesn't make a huge difference, though I found grouping by class more natural. E.g. I tend to rather think about all the attributes of a work package and not about all the classes that have an author_id. So from my side there's a slight difference towards classes as keys.

machisuji replied:
I mean I'd be fine with just a big table like

```
{
  "AuthProvider" => :creator_id,
  "Attachment" => :author_id,
  [...]
}.each do |model, column|
  Principals::ReplaceReferencesService.add_replacement model, column
end
```

too. You are right. It's probably a more sensible way to think about this with the models as the keys.

NobodysNightmare replied:
I added a "hash version" of the method and I agree that it looks nicer in the initializer.

Hoping that "looks" were the main motivation of your comment, I chose a simple implementation where I still call `add_replacement` a dozen times. This ensures that the passed data structure can't mess with the internal representation I wish to keep: Storing attributes in a set per class name.

I.e. I don't explicitly think about merging attribute lists and converting stuff back to sets. This happens in the original method.

What do you think?

edit:

> I mean I'd be fine with just a big table like

I read that after I wrote this comment... Yes. that's kind of what I did :D

machisuji replied:
Looks good to me!

machisuji reviewed (changes requested):
Looks good, thanks! I've just got two remarks.

NobodysNightmare committed:
Add short-hand syntaxes to add reference replacements

There is now add_replacements which accepts a hash, allowing to
add multiple replacements in one call. This should make mass-adding
replacements look less bulky.

Also including helpers to ActsAsOpEngine to make adding replacements
from a module nicer.

machisuji reviewed (approved):
👍
